doc/design/design_sqlflow_parser.md

```diff
@@ -0,0 +1,149 @@
+1. Support arbitrary select statements for extended syntax.
+
+The design of SQLFlow parser should support the above requirements.
+
+## Overview
+
+**The third-party parser** means the pre-existing parser of a particular SQL engine, like HiveQL/Calcite/TiDB parser. It supports parsing on a specific SQL dialect.
+
+**The extended syntax parser** means the parser written by the SQLFlow team to parse the extended syntax like `TO TRAIN/PREDICT/EXPLAIN ...`.
+
+**The SQLFlow parser** means the combination of the above two parsers. It supports parsing a SQL program of any SQL dialect.
+
+Upon receiving a SQL program like the following (`SELECT ...` represents a complicated select statement).
+
+```SQL
+CREATE TABLE my_training_table AS SELECT ...;
+
+SELECT ... TO TRAIN ...;
+
+SELECT ... TO PREDICT ...;
+
+SELECT * FROM predict_result_table;
+```
+
+The SQLFlow splits it into four statements.
+- For standard SQL statements like the `CREATE ...` and `SELECT * FROM predict_result_table`, SQLFlow calls the third party parser to check syntax.
+- For extended SQL statements like the second and the third statement, SQLFlow parser further splits the statement into two parts, namely a standard select statement and a train/predict statement.
+    - SQLFlow verifies the first half is a syntactically correct select statement.
+    - SQLFlow parses the second half using the extended parser.
+
+## Design Choices
+
+### Lexer vs. Third-Party Parser
+
+One major technical challenge is splitting, which includes
+
+1. Splitting a SQL program into multiple SQL statements.
+1. Splitting an extended SQL statement into a standard select statement and a train/predict statement.
+
+There are two possible solutions.
+
+#### Solution 1: Splitting using Lexer
+
+SQLFlow lexer scans the SQL program and receives a sequence of tokens.
+
+SQLFlow splits the SQL program by the `;` token.
+
+SQLFlow splits the extended SQL by looking for consecutive tokens returned by the lexer. If SQLFlow finds the consecutive tokens like [`TO`, `TRAIN`] or [`TO`, `PREDICT`], it splits the SQL string at the beginning of the first token in the list. For example, the lexer can go through the SQL statement `SELECT ... TO TRAIN` and find it satisfies the splitting criteria.
+
+Pros:
+1. Straight forward to implement.
+
+Cons:
+1. Incompleteness: single quote, double quote, square parentheses have different meanings in different SQL dialects. SQLFlow Lexer can't tokenize SQL programs of varying SQL dialects uniformly. This may lead to errors in the splitting.
+1. Needs to deal with special cases like `ALTER TABLE table_name RENAME TO train`.
+
+#### Solution 2: Splitting using the Third Party Parser
+
+SQLFlow begins parsing by using the third party parser(TPP). If TPP encounters an error at position `p`, TPP redoes parsing on the SQL statement before `p`, if success, extended SQL parser will parse on the statement after `p`. SQLFlow repeat these steps until the end of the SQL program.
```

Comment:
What to do if parsing failed before `p`?  Does it mean SQL statement is an invalid stander SQL?